aten/src/ATen/native/cuda/TensorModeKernel.cu

```diff
@@ -0,0 +1,322 @@
+    }
+  }
+}
+
+template <int64_t size, typename scalar_t>
+void handle_fused_mode(
+    dim3 grid,
+    const Tensor& self,
+    cuda::detail::TensorInfo<scalar_t, unsigned int>& ti_values,
+    cuda::detail::TensorInfo<int64_t, unsigned int>& ti_indices,
+    int64_t slice_size) {
+  const dim3 block(size / 2);
+  const auto memsize =
+      (sizeof(scalar_t) * size) + (2 * size * sizeof(unsigned int));
+  compute_mode<scalar_t, size>
+      <<<grid, block, memsize, at::cuda::getCurrentCUDAStream()>>>(
+          self.data_ptr<scalar_t>(), ti_values, ti_indices, slice_size);
+  C10_CUDA_KERNEL_LAUNCH_CHECK();
+}
+
+template <typename scalar_t>
+void fused_mode(
+    Tensor& values,
+    Tensor& indices,
+    const Tensor& self,
+    int64_t slice_size,
+    int64_t slices) {
+  // Set-up TensorInfo structs for passing to kernel
+  auto ti_values = cuda::detail::getTensorInfo<scalar_t, unsigned int>(values);
+  auto ti_indices = cuda::detail::getTensorInfo<int64_t, unsigned int>(indices);
+
+  // The number of blocks is the number of slices that we need to calculate
+  // the mode for. Each block is responsible for computing a single mode
+  dim3 grid;
+  getGridFromTiles(slices, grid);
+
+  // The blocksize is two elements per thread, rounded up to the nearest power
+  // of 2
+  auto ceilPowerOf2 = nextHighestPowerOf2(slice_size);
+
+  // Tradeoff between compilation time and the number of specializations.
+  // Ideally we would have one handle_fused_mode for each power of 2
+  switch (ceilPowerOf2) {
+    case 2048:
+      handle_fused_mode<2048, scalar_t>(
+          grid, self, ti_values, ti_indices, slice_size);
+      break;
+    case 1024:
+    case 512:
+    case 256:
+      handle_fused_mode<1024, scalar_t>(
+          grid, self, ti_values, ti_indices, slice_size);
+      break;
+    case 128:
+    case 64:
+      handle_fused_mode<128, scalar_t>(
+          grid, self, ti_values, ti_indices, slice_size);
+      break;
+    case 32:
+    case 16:
+    case 8:
+    case 4:
+    case 2:
+      handle_fused_mode<32, scalar_t>(
+          grid, self, ti_values, ti_indices, slice_size);
+      break;
+    case 1:
+    default:
+      TORCH_INTERNAL_ASSERT(false);
+  }
+
+  AT_CUDA_CHECK(cudaGetLastError());
+}
+
+void mode_kernel_impl(
+    Tensor& values,
+    Tensor& indices,
+    const Tensor& self,
+    int64_t dim,
+    bool keepdim) {
+  auto self_sizes = ensure_nonempty_vec(self.sizes().vec());
+  int64_t ndim = ensure_nonempty_dim(self.dim());
+  int64_t slice_size = ensure_nonempty_size(self, dim);
+  int64_t slices = self.numel() / slice_size;
+
+  // Resize output value, index Tensors to appropriate sizes (i.e. the same as
+  // the input Tensor, except at dim=dimension, the size is 1)
+  self_sizes[dim] = 1;
+
+  if (!keepdim) {
+    if (values.ndimension() >= dim) {
+      values.unsqueeze_(dim);
+    }
+    if (indices.ndimension() >= dim) {
+      indices.unsqueeze_(dim);
+    }
+  }
+
+  values.resize_(self_sizes);
```

Comment:
this would result in the wrong behavior if values and indices provided by user are discontiguous yet have correct size, see expected behavior https://github.com/pytorch/pytorch/wiki/Developer-FAQ#how-does-out-work-in-pytorch

Reply:
Do you mean that: in such a situation, `mode` would return contiguous `values` and `indices` instead of discontiguous values, right? Not sure I can see how that might happen, though... As far as I understand, given that the output tensors **have the correct size**, there are two cases:
1. `keepdims = True`: both output tensors have the same size as `self_sizes` (size 1 at `dim`)
2. `keepdims = False` (the scenario you are referring to): the condition `values.ndimension() >= dim` (same for `indices`) will always be true. Therefore, we will call `unsqueeze_` for them before `resize_`.

When `values` and `indices` have the correct size, `resize_` does nothing (AFAIK). Am I missing something? As it is now:

```python
> v.is_contiguous(), i.is_contiguous()
(False, False)
> torch.mode(a, dim = dim, keepdims = False, out = (v, i))
> v.is_contiguous(), i.is_contiguous()
(False, False)
```

Reply:
I believe @ysiraichi is correct, although this should probably be using `resize_output` instead of `resize_`. Note that `resize_output` unconditionally calls `resize_(shape)` internally, so this is entirely about the warning message.

https://github.com/pytorch/pytorch/blob/d54be1a9467db5075256d229ea1b01f1a4bcba8d/aten/src/ATen/native/Resize.cpp#L22

Also note that this `resize_` call is copied from existing code, so both need to be changed to use `resize_output`:
https://github.com/pytorch/pytorch/blob/d54be1a9467db5075256d229ea1b01f1a4bcba8d/aten/src/ATen/native/cpu/TensorCompareKernel.cpp#L36-L37

Reply:
Sorry, I didn't notice `unsqueeze_` above, the behavior is actually correct. @peterbell10 is right, resize_output should be called here.